File: ios/pqp/Sources/Voice/ServerMute.swift
import Foundation

/// A moderator's mute, as both call models and both screens read it.
///
/// WHY THIS IS A PURE TYPE. `VoiceModel` and `CallModel` each own a room, and
/// the rules below would otherwise be written twice by eye and drift the way
/// `callStageLayout` nearly did before it was pulled out. Free of SwiftUI and
/// WebRTC so the unit-test target, which cannot open a microphone, can pin them.
///
/// THE CONTRACT, shared with the web and Android clients. `VoiceParticipant`
/// carries `serverMuted`, defaulting to false when an older server omits it,
/// on every frame that carries a participant: `welcome`, `voice-roster`,
/// `peer-joined`, `peer-updated`. On mesh the server never touches media, so
/// the flag is only real because each receiver plays that person at zero (see
/// `RemoteAudioMixer.setServerMuted`) and the target's own client stops the
/// mic. The server refuses the target's `set-muted false` and snaps the roster
/// back, which is why the control goes off rather than merely red.
enum ServerMute {
    /// Whether the mic button does anything.
    ///
    /// `connected` is the room's own gate (a voice channel disables every
    /// control until `welcome`; a DM call never does). The moderator's flag is
    /// the only thing that turns the button off in a connected room.
    static func muteControlIsEnabled(connected: Bool, selfServerMuted: Bool) -> Bool {
        connected && !selfServerMuted
    }

    /// What our own microphone becomes when the roster describes us.
    ///
    /// Muted by a moderator means muted, whatever we had chosen: our client is
    /// the only thing on mesh that can stop the bytes leaving. The flag
    /// CLEARING changes nothing here on purpose, which is why it is not an
    /// argument that can flip the result back: coming back live the instant a
    /// moderator lets go would put whatever is being said mid-sentence into
    /// the room. Unmuting is the person's own tap.
    static func selfMuted(currentlyMuted: Bool, serverMuted: Bool) -> Bool {
        currentlyMuted || serverMuted
    }

    /// The glyph for a moderator's mute. Deliberately not `mic.slash.fill`,
    /// which is what a person who muted themselves shows: one is a choice the
    /// person can undo and the other is something done to them, and a room
    /// that draws them the same cannot tell "they stepped away" from "they
    /// were silenced".
    static let glyph = "mic.slash.circle.fill"

    /// The self-muted glyph, named here so the two are visibly a pair and a
    /// change to one is made next to the other.
    static let selfMutedGlyph = "mic.slash.fill"

    /// The one sentence the target sees.
    static var notice: String {
        String(localized: "A moderator muted you. You can talk again when they unmute you.")
    }
}

File: android/app/src/main/kotlin/gg/pqp/app/voice/ServerMute.kt
package gg.pqp.app.voice

import gg.pqp.app.core.VoiceParticipant

/**
 * What a roster says about a moderator having muted *this* device.
 *
 * Pulled out of [VoiceController] so it can run on the JVM: the controller
 * needs a `Context` and an engine that needs libwebrtc, and the rule below is
 * the one piece of the server-mute contract about the local person rather than
 * about a peer, so it deserves a test that does not need either.
 *
 * The rule: find ourselves on the roster by peer id, copy the flag, and while it
 * is set force the local mute on. The server rejects `set-voice-state` with
 * `muted: false` from a server-muted peer and snaps the roster back, so a client
 * that kept showing a live microphone would be lying for a round trip and then
 * flickering. Clearing the flag does **not** unmute: the person was muted by
 * somebody else, and handing their microphone back without asking is the one
 * thing a mute control must never do. They unmute themselves, as on the web.
 */
internal fun VoiceState.absorbServerMute(
    participants: List<VoiceParticipant>,
    localPeerId: String?,
): VoiceState {
    val self = participants.firstOrNull { it.peerId == localPeerId }
    val serverMuted = self?.serverMuted ?: this.serverMuted
    return copy(
        serverMuted = serverMuted,
        muted = muted || serverMuted,
    )
}

/**
 * Whether the mute button does anything right now.
 *
 * One expression, named, so the controller and the call bar cannot disagree
 * about it: the button is drawn inert for exactly the state in which a tap is
 * ignored, and the test pins both to the same fact.
 */
internal val VoiceState.muteControlEnabled: Boolean
    get() = !serverMuted

/**
 * Everybody on the roster a moderator has muted, other than this device.
 *
 * The local person is left out because their case is drawn differently (the
 * mute control itself, and the status line), not because the flag means
 * something else for them.
 */
internal fun VoiceState.serverMutedPeers(): List<VoiceParticipant> =
    participants.filter { it.serverMuted && it.peerId != localPeerId }
